File: VoigtKampff/VoigtKampff.cpp
#include <math.h>
#include <vector>
#include <iostream>
#include <algorithm>
#include "VoigtKampff.h"
#include <cstdlib>
#include <cstdio>
const double PI = 3.14159265359;
const double SQRTLN2 = 0.832554611;
const double ISQRTPI = 0.564189584; 

const double DISTANCE_MAGIC_NUMBER = 1.9;
const double REFERNECE_NU = 1.0;
double CubicInterpolate(
   double y0,double y1,
   double y2,double y3,
   double mu)
{
   double a0,a1,a2,a3,mu2;

   mu2 = mu*mu;
   a0 = y3 - y2 - y0 + y1;
   a1 = y0 - y1 - a0;
   a2 = y2 - y0;
   a3 = y1;

   return(a0*mu*mu2+a1*mu2+a2*mu+a3);
};


VoigtKampff::VoigtKampff(double pGammaD,double pRes,double pLorenzCutoff,bool pNormalize){
	m_Npoints = 2*pLorenzCutoff/pRes + 1;
	middle_point = m_Npoints/2;
	m_res=pRes;
	m_gammaD = pGammaD;
	m_lorentz_cutoff = pLorenzCutoff;
	m_normalize = pNormalize;
	num_gammaL=0;


}


VoigtKampff::VoigtKampff(std::vector<double> pGammaL,double pGammaD,double pRes,double pLorenzCutoff,bool pNormalize) : VoigtKampff(pGammaD,pRes,pLorenzCutoff,pNormalize){
	int dummy;
	//num_gammaL=0;
	//m_Npoints = 2*pLorenzCutoff/pRes + 1;
	//middle_point = m_Npoints/2;
	//m_res=pRes;
	//m_gammaD = pGammaD;
	//m_lorentz_cutoff = pLorenzCutoff;
	
	std::cout<<"Initializing Voigt....."<<std::endl;
	std::cout<<"Npoints: "<<m_Npoints<<" middle: "<<middle_point<<"Humlecik points: "<<DISTANCE_MAGIC_NUMBER/m_res<<" range : "<< 0.0 << " - "<< double(m_Npoints)*m_res<<std::endl;
	m_gammaL = pGammaL;
	double x,y;
	double nu = 0.0;
	for(int i = 0; i < pGammaL.size(); i++){
		double gammaL = pGammaL[i];
		AddLorentzian(gammaL,&dummy);
	}

	std::cout<<"Done!!!"<<std::endl;



}

int VoigtKampff::CheckLorentzian(double gammaL){
			int gammaIdx=-1;
			std::map<float,int>::iterator i = m_gamma_mapping.find((float)gammaL);
			if(i == m_gamma_mapping.end()){
				gammaIdx = -1;
				//AddLorentzian(gammaL);
			}else{
				gammaIdx = i->second;
			}	
			return gammaIdx;
}


double VoigtKampff::ComputeVoigt(double dfreq,int gammaL,double nu){
	
	double dfreq_=fabs(dfreq);
	if(dfreq_ > m_lorentz_cutoff)
		return 0.0;
	int point = dfreq/m_res + middle_point;
	//printf("point:%d\n",point);
	//point = std::max(point,0c);
	//point = std::min(point,m_Npoints);
	
		
		//TODO: Use intepolation 
		/*int i1 = std::max(point-2,0);
		int i2 = std::max(point-1,0);
		int i3 =  std::min(point+2,m_Npoints);
		int i4 = std::min(point+2,m_Npoints);
		double p1=m_voigt_grid[gammaL][i1];
		double p2=m_voigt_grid[gammaL][i2];
		double p3=m_voigt_grid[gammaL][i3];
		double p4=m_voigt_grid[gammaL][i4];
		return CubicInterpolate(p1,p2,p3,p4,dfreq);///ComputeDoppler(dfreq,m_gammaD,nu);
		*/
	//double dist = 2.0*m_gammaD*nu/m_gammaL[gammaL];
	//if(dfreq_ < dist) return HumlicekTest(dfreq,m_gammaL[gammaL],nu);
	return m_voigt_grid.at(gammaL).at(point);
		//;
	//return 0.0;
	
}	

void VoigtKampff::AddLorentzian(double gammaL,int* index){
		double nu = -m_lorentz_cutoff;
		*index = m_voigt_grid.size();
		m_voigt_grid.push_back(std::vector<double>());	
		
		m_gamma_mapping[(float)gammaL]= num_gammaL;
		m_gammaL.push_back(gammaL);
		m_mag.push_back(0.0);
		for(int j = 0; j < m_Npoints; j++){	
			
			double x = SQRTLN2*fabs(nu)/(m_gammaD*REFERNECE_NU);
			double y = SQRTLN2*gammaL/(m_gammaD*REFERNECE_NU);
		
			m_voigt_grid.back().push_back(humlic(x,y)*SQRTLN2*ISQRTPI/(m_gammaD*REFERNECE_NU));
			m_mag.back() += m_voigt_grid.back().back();
			nu+=m_res;
		}
		std::cout<<"mag:"<< m_mag.back()*m_res<<std::endl;
		m_mag.back() *= m_res;
		num_gammaL++;
}


void VoigtKampff::DoVectorized(double* __restrict intens,const std::vector<double> & gammaVoigt,const double abscoef,const int start,const int end){
	//int count = end-start;
	
	for(int i = start; i < end; i++){
		intens[i]+=gammaVoigt[i]*abscoef;
	}


}

void VoigtKampff::ComputeVoigtVectorized(const double* __restrict freq,double* __restrict intens,const double abscoef,const int ib,const int ie,const double gammaL,const double start_nu,const double nu){
			int gammaIdx = CheckLorentzian(gammaL);
			//printf("%d\n",gammaIdx);


			ComputeVoigtVectorized(freq, intens,abscoef,ib,ie,gammaIdx,start_nu,nu);
}

void VoigtKampff::ComputeVoigtVectorized(const double* __restrict freq,double* __restrict intens,const double abscoef,const int ib,const int ie,const int gammaL,const double start_nu,const double nu){


	//int ib=ib_-1;
	int center_point = (nu-start_nu)/m_res;
	int middle_shift = center_point - middle_point;
	double gammaL_ = m_gammaL[gammaL];
	
	//int dist = DISTANCE_MAGIC_NUMBER/m_res;//(m_gammaD*nu/m_gammaL[gammaL])/m_res;
	int dist;
	if (gammaL_ == 0.0)
		dist = m_Npoints;
	else
		//dist = std::max(m_gammaD / m_res,2.0);
		//dist = std::max((m_gammaD*nu/m_gammaL[gammaL])/m_res,8.0);
		dist = std::max(2.0/m_res,3.0);
	int ib_rel = ib - middle_shift;
	int ie_rel = ie - middle_shift;
	
	double gammaD_ = (m_gammaD*nu);

	
	//dist = std::max(101,dist);
	//int start_dist 
	//printf("Middle calc %d %d\n",start_dist,end_dist);
	
	
	
	double dfreq;
	//DO PURE DOPPLER
	if (gammaL_ == 0.0) {
		//USe vectorized doppler
		DoDopplerVectorized(freq, intens, abscoef, ib, ie, gammaD_, nu);



	}//OTHERWISE WE VOIGT BABEEEEEEEEEEEEEEE
	else {
		int Npoints = ie - ib + 1;
		double mag = m_mag[gammaL];

		if(!m_normalize) mag = 1.0;
		
		int begin_humlicek;
		int end_humlicek;
		int begin_reflect;
		int end_reflect;
		int fudge_point;
		
		//Do the middle
		int start_dist = std::max(center_point - dist, ib);
		int end_dist = std::min(center_point + dist, ie);
		
		int left_start = ib_rel;
		int left_end = std::min(middle_point - dist, ie_rel);
		int right_start = std::max(middle_point + dist, ib_rel);
		int right_end = ie_rel;
		
		if((left_end - left_start) >= (right_start - right_end)){
			begin_humlicek =start_dist;
			end_humlicek=std::min(center_point+2,ie);
			fudge_point = start_dist;
			begin_reflect=end_humlicek;
			end_reflect = end_dist;
			
		}else{
			begin_humlicek =std::max(center_point-1,start_dist);
			end_humlicek=end_dist;
			fudge_point = end_humlicek;
			begin_reflect=start_dist;
			end_reflect =begin_humlicek;
		}
		
		int total_hum_points;
		//printf("Middle calc %d %d %d %d %d %d %d %12.6f\n",begin_humlicek,center_point,end_humlicek,begin_reflect,end_reflect,ib,ie,freq[center_point]);
		
		double fudge_factor=1.0f;
		/*if (ib < center_point) {
			if(left_end >= left_start && left_end >0){
			 	fudge_factor =  HumlicekTest(freq[ib - ib_rel+left_end-1] - nu, gammaL_, nu)/m_voigt_grid[gammaL][left_end-1];
			 }

		
		}else if (ie >= center_point){
			if(right_end >= right_start && right_start >0){
				fudge_factor =  HumlicekTest(freq[ib - ib_rel+right_start-1] - nu, gammaL_, nu)/m_voigt_grid[gammaL][right_start-1];
			}
	
		}*/
		
		//mag /= fudge_factor;
		
		double* temp_humlicek;
		if (start_dist < end_dist) {
			total_hum_points = end_dist - start_dist + 1;
			temp_humlicek = new double[total_hum_points];

			for (int i = start_dist; i < end_dist; i++) {
				dfreq = freq[i] - nu;
				double hum = HumlicekTest(dfreq, gammaL_, nu);
				temp_humlicek[i - start_dist] = hum;
				mag -= m_voigt_grid[gammaL][i -middle_shift]*m_res;
				mag += hum*m_res;
			}
			
			//fudge_factor = temp_humlicek[fudge_point - start_dist]/m_voigt_grid[gammaL][fudge_point -middle_shift];
			fudge_factor = 1.0;
			//This is so that the second one is vectorized
			if(!m_normalize) mag = 1.0;
			
			for(int i = start_dist; i < end_dist; i++){
				intens[i] += temp_humlicek[i-start_dist]*abscoef/mag;
				
				
			}
			
			delete[] temp_humlicek;

		}
		//printf("ib: %d ie: %d start:%d end:%d dist: %d center:%d nu: %12.6f\n",ib,ie,ib_rel,ie_rel,dist,center_point,nu);

		//If we have a calculation on the left
		if (ib < center_point) {
			DoVectorized(intens + ib - ib_rel, m_voigt_grid[gammaL], (abscoef/mag)*fudge_factor, left_start, left_end);
			//printf("Left calc %d %d\n",left_start,left_end);

		}
		if (ie >= center_point) {

			DoVectorized(intens + ib - ib_rel, m_voigt_grid[gammaL], (abscoef/mag)*fudge_factor, right_start, right_end);
			//printf("Right calc %d %d\n",right_start,right_end);

		}


	}
	


	//int dist_hum = dist;
	
	//for(int i = dist+1; i < dist
	
	
	
	//exit(0);
	
	
	
	



}

void VoigtKampff::DoDopplerVectorized(const double * __restrict freq, double * __restrict intens, const double abscoef, const int ib, const int ie, const double gammaD, const double nu)
{
	double fact = SQRTLN2 / gammaD;
	double x0 = fact*m_res*0.5;
	
	for (int i = ib; i < ie; i++) {
	//	double dfreq = freq[i] - nu;
		//double xp = fact*(dfreq)+x0;
		//double xm = fact*(dfreq)-x0;
	//	double de = erf(xp) - erf(xm);
		intens[i] += abscoef*0.5/m_res*(erf(fact*(freq[i] - nu) + x0)
			- erf(fact*(freq[i] - nu) - x0));
			//intens[i] += exp(i);
			/*intens[i] +=
				//DE
				( erf(   fact*(freq[i] - nu) + x0   )
				- erf(  fact*(freq[i] - nu) - x0  ) )

				*abscoef*0.5 /
				(freq[i] - nu);
		}*/
	}
}


double VoigtKampff::ComputeDoppler(double dfreq,double gammaD,double nu){
	
    	 //double dop = 0.03912;//(3155.87/nu)*4.523*(gammaD*nu);
         /*  
         double xp = SQRTLN2/dop*(freq+ m_res/2.0 - nu);
         double xm = SQRTLN2/dop*(freq- m_res/2.0 - nu);
             
         double de = erf(xp)-erf(xm);
        
         return 0.5/m_res*de;
	*/
	
	//double alpha = -1.0/(dop);
	//double de = exp(alpha*dfreq_*dfreq_);
	//double de = (dfreq*dfreq + (dop)*(dop))/(dop*dop*nu) ;
	//return de;
	return (-3.295e-05*dfreq*dfreq*dfreq*dfreq + 2.916e-08*dfreq*dfreq*dfreq 
		+ 0.2924*dfreq*dfreq -6.335e-06*dfreq + 0.02171);


}


	//void Initialize()
double VoigtKampff::ComputeVoigt(double dfreq,double gammaL){
	
	double dfreq_=fabs(dfreq);
	if(dfreq_ > m_lorentz_cutoff)
		return 0.0;
	int point = fabs(dfreq_)/m_res;
	//point = std::max(point,0);
	
	
	if(m_gamma_mapping.count(gammaL) > 0){
		int gamma_idx = m_gamma_mapping[gammaL];
		
		//TODO: Use intepolation 
		/*int i1 = std::max(point-1,0);
		int i2 = point;
		int i3 =  std::min(point+2,m_Npoints);
		int i4 = std::min(point+2,m_Npoints);
		double p1=m_voigt_grid[gamma_idx][i1];
		double p2=m_voigt_grid[gamma_idx][i2];
		double p3=m_voigt_grid[gamma_idx][i3];
		double p4=m_voigt_grid[gamma_idx][i4];
		return CubicInterpolate(p1,p2,p3,p4,dfreq);
		*/
		
		return m_voigt_grid[gamma_idx][point];
		//;
	}
	return 0.0;
	
}



double VoigtKampff::HumlicekTest(double dfreq,double gammaL,double nu){
	double x,y;
	double dfreq_=fabs(dfreq);
	double gammaD = m_gammaD*nu;
	
	x = SQRTLN2*dfreq_/(gammaD);
	y = SQRTLN2*gammaL/gammaD;
	return humlic(x,y)*SQRTLN2*ISQRTPI/gammaD;
}

File: VoigtKampff.f90
module VoigtKampffCollection_module
    use accuracy
    implicit none

    public 	VoigtKampffCollection

    private

    integer(ik),parameter :: EXPAND_SIZE = 50

    real(rk),parameter    ::    REFERENCE_NU0 = 1.0
    real(rk),parameter    ::    DISTANCE_MAGIC_NUMBER = 2.0

    type ::        VoigtKampff
        private
        integer(ik)    ::    m_Npoints
        real(rk)    ::    m_res
        real(rk)    ::    m_lorentz_cutoff
        integer(ik)    ::    m_middle_point
        real(rk)     ::    m_gammaD
        real(rk)     ::    m_gammaL
        real(rk)    ::    m_mag
        logical        ::    normalize
        logical        ::    constructed = .false.
        real(rk),pointer    ::    m_voigt_grid(:)


    contains
        procedure,public    ::    construct    =>    construct_voigt
    
        procedure,public    ::    compute        =>     compute_voigt
    
        procedure,public        ::    destroy

        procedure,public    ::    get_gammaL

    end type    
    
    
    
    type ::	 VoigtKampffCollection

        type(VoigtKampff),allocatable :: fast_voigts(:)
        integer(ik) :: capacity
        integer(ik) :: n
        logical     :: normalize
        real(rk)    :: dpwcoeff
        real(rk)    :: offset
        real(rk)    :: dfreq

    contains
        procedure,public  :: construct
        procedure,public  :: expand_voigts
        procedure,public  :: generate_indices
        procedure,public  :: get_size
        procedure,private :: add_voigt
        procedure,private :: search_gamma
        procedure,public  :: compute => compute_fast_voigt
    end type

    
    

contains

    subroutine construct(this,dpwcoeff,offset,dfreq,normalize)
        class(VoigtKampffCollection)	::	this
        logical,intent(in)		::	normalize
        real(rk),intent(in)		::	dpwcoeff,offset,dfreq

        this%n = 0
        this%capacity = 0

        this%dpwcoeff =dpwcoeff
        this%offset = offset
        this%dfreq = dfreq
        this%normalize = normalize

        call this%expand_voigts()

    end subroutine

    subroutine expand_voigts(this)
        class(VoigtKampffCollection) :: this
        type(VoigtKampff),allocatable :: temp(:)
        integer(ik) :: old_capacity

        if(this%capacity == 0) then
            this%capacity = this%capacity + EXPAND_SIZE
            allocate(this%fast_voigts(this%capacity))
            return
        endif

        allocate(temp(this%capacity))
        temp(:) = this%fast_voigts(:)

        deallocate(this%fast_voigts)

        old_capacity=this%capacity

        this%capacity = this%capacity + EXPAND_SIZE

        allocate(this%fast_voigts(this%capacity))

        this%fast_voigts(1:old_capacity) = temp(1:old_capacity)

        deallocate(temp)

    end subroutine

    integer(ik) function get_size(this)
        class(VoigtKampffCollection),intent(in)	::	this

        get_size = this%n
    end function


    subroutine generate_indices(this,gammaL,gamma_idx,Jmax)
        class(VoigtKampffCollection) :: this
        real(rk),intent(in) :: gammaL
        integer(ik),intent(inout) :: gamma_idx
        integer(ik),intent(in) :: Jmax

        integer(ik) :: ido,jdo
        real(rk)    :: val
        integer     :: idx

        !Search of the gamma value, if its similar then resue the index otherwise we add antoher voigt

        val = gammaL
        idx = this%search_gamma(val)
              !If we didnt find it then add a voigt
        if(idx==0) then
           gamma_idx = this%add_voigt(val)
        else
           gamma_idx = idx
        endif

    end subroutine

    subroutine compute_fast_voigt(this,freq,intens,abscoef,ib,ie,start_nu,nu0,index)
        class(VoigtKampffCollection),intent(in)	::	this
        real(rk),intent(in)     :: freq(:),abscoef,start_nu,nu0
        integer(ik),intent(in)  :: ib,ie
        real(rk),intent(inout)  :: intens(:)
        integer,intent(in)      :: index
        if(index > this%n) stop "FastVoigt::Out of bounds"

        call this%fast_voigts(index)%compute(freq,intens,abscoef,ib,ie,start_nu,nu0)

    end subroutine

    integer(ik) function add_voigt(this,gammaL)
        class(VoigtKampffCollection) :: this
        real(rk),intent(in) :: gammaL

        this%n = this%n+1

        if(this%n > this%capacity) call this%expand_voigts

        add_voigt = this%n

        call this%fast_voigts(this%n)%construct(this%dpwcoeff,gammaL,this%dfreq,this%offset,this%normalize)
   
    end function

    integer(ik) function search_gamma(this,gammaL)
        class(VoigtKampffCollection) :: this
        real(rk),intent(in) :: gammaL

        search_gamma = 0

        if(this%n == 0) return
        search_gamma = binarySearch_R(this%fast_voigts(1:this%n),gammaL)


    end function

recursive function binarySearch_R (a, value) result (bsresult)
    type(VoigtKampff), intent(in) :: a(:)
    real(rk)	::	value
    integer          :: bsresult, mid
 
    mid = size(a)/2 + 1
    if (size(a) == 0) then
        bsresult = 0        ! not found
    else if(is_close(a(mid)%get_gammaL(),value,1e-5_rk)) then
        bsresult = mid      ! SUCCESS!!
    else if (a(mid)%get_gammaL() > value) then
        bsresult= binarySearch_R(a(:mid-1), value)
    else if (a(mid)%get_gammaL() < value) then
        bsresult = binarySearch_R(a(mid+1:), value)
        if (bsresult /= 0) then
            bsresult = mid + bsresult
        end if
   endif
end function binarySearch_R



    !Check if the lorentzian is the same
    logical function is_close(a, b, rel_tol )
      real(rk),intent(in)	::	a,b,rel_tol
      	is_close = abs(a-b) <= rel_tol * max(abs(a), abs(b))
    end function



  subroutine construct_voigt(this,pGammaD,pGammaL,pRes,pLorentzCutoff,pNormalize)
      class(VoigtKampff),intent(inout)    ::    this
      real(rk),intent(in)            ::    pGammaD,pGammaL
      real(rk),intent(in)            ::    pRes
      real(rk),intent(in)            ::    pLorentzCutoff
      logical,intent(in)            ::    pNormalize

      integer(ik)                ::    ierr,ido

      real(rk)                ::    nu,nu0

  
      this%m_res = pRes
      this%m_gammaD = pGammaD
      this%m_gammaL = pGammaL
      this%m_lorentz_cutoff = pLorentzCutoff
     this%normalize = pNormalize
      this%m_Npoints = int(2.0*this%m_lorentz_cutoff/this%m_res,ik);
      this%m_middle_point = (this%m_Npoints/2) + 1
      !Construct the voigt grid

      this%m_mag = 0.0

      allocate(this%m_voigt_grid(this%m_Npoints),stat=ierr)

      !call ArrayStart('Voigt-F-Grid',info,size(this%m_voigt_grid),kind(this%m_voigt_grid))
      nu0 = REFERENCE_NU0
      nu = nu0-this%m_lorentz_cutoff

      do ido=1,this%m_Npoints
      
          this%m_voigt_grid(ido) = voigt_humlicek(nu,nu0,this%m_gammaD,this%m_gammaL)
          this%m_mag = this%m_mag + this%m_voigt_grid(ido)
          nu = nu+ this%m_res
      enddo

      this%m_mag = this%m_mag*this%m_res

      this%constructed = .true.

  end subroutine

  subroutine destroy(this)
     class(VoigtKampff),intent(inout)    ::    this
     
     if(.not. this%constructed)return
     
     if(associated(this%m_voigt_grid)) then
         deallocate(this%m_voigt_grid)
     endif
  end subroutine

  
  subroutine compute_voigt(this,freq,intens,abscoef,ib,ie,start_nu,nu0)
      class(VoigtKampff),intent(in)    ::    this
      real(rk),intent(in)        ::    freq(:),abscoef,start_nu,nu0
      integer(ik),intent(in)        ::    ib,ie
      real(rk),intent(inout)        ::    intens(:)

  
      integer                ::    center_point,dist
      integer                ::    middle_shift
      integer                ::    start_dist,end_dist,left_start,left_end,right_start,right_end
      integer                ::    ib_rel,ie_rel,num_hum_points,ido
      integer                ::    Npoints
      real(rk)            ::    gammaD,hum_res,mag,nu
      real(rk),allocatable        ::    temp_humlicek(:)

  
      center_point = (nu0-start_nu)/this%m_res + 1
      middle_shift = (center_point - this%m_middle_point) + 1

  

      dist = max(DISTANCE_MAGIC_NUMBER/this%m_res,3.0)

      Npoints = ie - ib + 1

      ib_rel = ib - middle_shift
    ie_rel = ie - middle_shift

      gammaD = this%m_gammaD*nu0


    mag = this%m_mag

    start_dist = max(center_point - dist + 1, ib)
    end_dist = min(center_point + dist, ie)
    
    left_start = ib_rel
    left_end = min(this%m_middle_point - dist, ie_rel)
    right_start = max(this%m_middle_point + dist, ib_rel)
    right_end = ie_rel

    !print *,mag


    !If we have any humlicek points
    if(start_dist < end_dist) then


        if(this%normalize) then 
            num_hum_points = end_dist - start_dist + 1;
            allocate(temp_humlicek(num_hum_points))

            do ido=start_dist,end_dist
                nu=freq(ido)
                hum_res = voigt_humlicek(nu,nu0,gammaD,this%m_gammaL)
                temp_humlicek(ido-start_dist+1) = hum_res
                mag = mag - this%m_voigt_grid(ido -middle_shift+1)*this%m_res
                mag = mag + hum_res*this%m_res
            enddo
    
    
            intens(start_dist:end_dist) = intens(start_dist:end_dist) + temp_humlicek(:)*abscoef/mag
    
            !print *,mag
            deallocate(temp_humlicek)
        else
            do ido=start_dist,end_dist
                nu=freq(ido)
                intens(ido) = intens(ido) + voigt_humlicek(nu,nu0,gammaD,this%m_gammaL)*abscoef
            enddo    
        
            mag = 1.0
    
    
        endif
      endif
      if (left_start < left_end ) then
          call vectorized_voigt(intens(ib-ib_rel+left_start:ib-ib_rel+left_end),this%m_voigt_grid(left_start:left_end),(abscoef/mag))

      endif

      if (right_start < right_end ) then
          call vectorized_voigt(intens(ib-ib_rel+right_start:ib-ib_rel+right_end),this%m_voigt_grid(right_start:right_end),(abscoef/mag))
      endif

  end subroutine
          
  real(rk) function get_gammaL(this)
      class(VoigtKampff),intent(in)    ::    this

      get_gammaL = this%m_gammaL

      return
  end function




!-----------Humlicek methods----------------------!

 ! This should be heavily vectorized by the compiler
  subroutine vectorized_voigt(intens,voigt,abscoeff)
      real(rk),intent(inout)    ::    intens(:)
      real(rk),intent(in)    ::    voigt(:)
      real(rk),intent(in)    ::    abscoeff

      intens(:) = intens(:)+voigt(:)*abscoeff
  end subroutine


  function voigt_humlicek(nu,nu0,alpha,gamma) result(f)
   !
   real(rk),intent(in) :: nu,nu0,alpha,gamma
   real(rk) :: f,sigma,x,y

   real(rk),parameter ::  RT2LN2 = 1.1774100225154747_rk     ! sqrt(2ln2)
   real(rk),parameter ::  RTPI   = 1.7724538509055159_rk     ! sqrt(pi)
   real(rk),parameter ::  RT2PI  = 2.5066282746310002_rk     ! sqrt(2pi)
   real(rk),parameter ::  RT2    = 1.4142135623730951_rk     ! sqrt(pi)
   !
   sigma = alpha / RT2LN2
   x = (nu0 - nu) / sigma / RT2
   y = gamma / sigma / RT2
   !
   f = humlic(x, y) / sigma / RT2PI
   !
  end function voigt_humlicek

  !
  function humlic(x,y) result (f)
    real(rk),intent(in) :: x,y
    real(rk) :: f
    complex(rk) :: t,humlic1,u
    real(rk) :: s
    !
    T = cmplx(y,-x)
    S = abs(x) + y
    !
    if (S >= 15) then
        ! Region I
        humlic1 = T*0.5641896_rk/(0.5_rk+T*T)
        !fprintf(stdout, "I")
        !
        f = real(humlic1,rk)
        return
    endif
    !
    if (S >= 5.5) then
        ! Region II
        U = T * T;
        humlic1 = T * (1.410474_rk + U*0.5641896_rk)/(0.75_rk + U*(3.0_rk+U))
        !
        !fprintf(stdout, "II");
        !
        f = real(humlic1,rk)
        return
        !
    endif
    !
    if (y >= 0.195_rk * abs(x) - 0.176_rk) then
        ! Region III
        humlic1 = (16.4955_rk+T*(20.20933_rk+T*(11.96482_rk &
                +T*(3.778987_rk+T*.5642236_rk)))) / (16.4955_rk+T*(38.82363_rk &
                +T*(39.27121_rk+T*(21.69274_rk+T*(6.699398_rk+T)))))
        !fprintf(stdout, "III");
        f = real(humlic1,rk)
        return
        !
    endif
    !
    ! Region IV
    U = T * T
    humlic1 = exp(U)-T*(36183.31_rk-U*(3321.9905_rk-U*(1540.787_rk-U*(219.0313_rk-U* &
       (35.76683_rk-U*(1.320522_rk-U*.56419_rk))))))/(32066.6_rk-U*(24322.84_rk-U* &
       (9022.228_rk-U*(2186.181_rk-U*(364.2191_rk-U*(61.57037_rk-U*(1.841439_rk-U)))))))
    !fprintf(stdout, "IV");
    !
    f = real(humlic1,rk)
    !
    end function humlic




end module VoigtKampffCollection_module
